crates/aptos/src/update/prover_dependency.rs

```diff
@@ -0,0 +1,363 @@
+                        binary_name, version, version_output
+                    ))
+                }
+            },
+            Err(e) => Err(format!(
+                "Failed to run {}: {}",
+                install_path.to_string_lossy(),
+                e
+            )),
+        }
+    }
+
+    async fn download_dependency(&self) -> CliTypedResult<String> {
+        let arch = get_arch();
+        let build_info = cli_build_information();
+        let target_res = match build_info.get(BUILD_OS).context("Failed to determine build info of current CLI")?.as_str() {
+            "linux-x86_64" => Ok("linux"),
+            "macos-aarch64" | "macos-x86_64" => Ok("macos"),
+            "windows-x86_64" => Ok("win"),
+            wildcard => Err(CliError::UnexpectedError(format!("Self-updating is not supported on your OS ({}) right now, please download the binary manually", wildcard))),
+        };
+
+        let target = target_res?;
+
+        let install_dir = match self.install_dir.clone() {
+            Some(dir) => dir,
+            None => {
+                let dir = get_additional_binaries_dir();
+                // Make the directory if it doesn't already exist.
+                std::fs::create_dir_all(&dir)
+                    .with_context(|| format!("Failed to create directory: {:?}", dir))?;
+                dir
+            },
+        };
+
+        let dir_name = format!("{}-{}", target, arch);
+
+        BoogieOptions::check_version_is_compatible(
+            BOOGIE_BINARY_NAME,
+            TARGET_BOOGIE_VERSION,
+            MIN_BOOGIE_VERSION,
+            MAX_BOOGIE_VERSION,
+        )?;
+        self.install_binary(
+            &install_dir,
+            BOOGIE_EXE,
+            BOOGIE_BINARY_NAME,
+            TARGET_BOOGIE_VERSION,
+            "/version",
+            BOOGIE_EXE_ENV,
+            &dir_name,
+        )
+        .await?;
+
+        BoogieOptions::check_version_is_compatible(
+            Z3_BINARY_NAME,
+            TARGET_Z3_VERSION,
+            MIN_Z3_VERSION,
+            MAX_Z3_VERSION,
+        )?;
+        self.install_binary(
+            &install_dir,
+            Z3_EXE,
+            Z3_BINARY_NAME,
+            TARGET_Z3_VERSION,
+            "--version",
+            Z3_EXE_ENV,
+            &dir_name,
+        )
+        .await?;
+
+        #[cfg(unix)]
+        {
+            BoogieOptions::check_version_is_compatible(
+                CVC5_BINARY_NAME,
+                TARGET_CVC5_VERSION,
+                MIN_CVC5_VERSION,
+                MAX_CVC5_VERSION,
+            )?;
+            self.install_binary(
+                &install_dir,
+                CVC5_EXE,
+                CVC5_BINARY_NAME,
+                TARGET_CVC5_VERSION,
+                "--version",
+                CVC5_EXE_ENV,
+                &dir_name,
+            )
+            .await?;
+        }
+
+        Ok("Installation succeeded".to_string())
+    }
+
+    async fn install_binary(
```

Comment:
The self update crate that we use in all the other `aptos update` subcommands does exactly this minus the env var stuff, any reason we can't use that?

Reply:
refactored the code to use the same mechanism